### ISSUE TITLE
Spurious task polling when task contains I/O-based stream

### ISSUE THREAD

**Issue by jonhoo** [tokio-rs/tokio#2222]
## Version

`master` as of 55b5e1b6adcc3f85a35d16444e496b203648c99d

## Platform

```
Linux xpanse 5.5.2-arch1-1 #1 SMP PREEMPT Tue, 04 Feb 2020 18:56:18 +0000 x86_64 GNU/Linux
```

## Description

When using a `timerfd`-based [implementation](https://github.com/DoumanAsh/async-timer) of an interval `Stream`, I noticed that my top-level task would frequently get polled one extra time every time it yielded an item: https://github.com/DoumanAsh/async-timer/issues/9. Concretely what I was seeing was:

```
poll_next() -> Pending
# time passes according to the interval time
poll_next() -> Ready(Some)
poll_next() -> Pending
poll_next() -> Pending
```

Notice that there is an unnecessary extra call to `poll_next` at the end — the moment we return `Pending` from `poll_next`, we should (in theory) not be polled again until the file descriptor is ready (i.e., once the interval time has passed). That last call to `poll_next` checks the file descriptor, and it is not ready (as expected).

Digging some more revealed that not only was `Stream::poll_next` called an extra time, but the whole _task_ was being polled again. The poll tree looked something like:

```
task.poll()
  interval.poll_next() -> Ready(Some)
  interval.poll_next() -> Pending
task.poll()
  interval.poll_next() -> Pending
```

This can be really bad for users with large tasks, as such a spurious wakeup may cause [a lot of extra work](https://github.com/mit-pdos/noria/blob/20b67f87fa56449d91bf0fabf0915232a394ad97/noria-server/src/worker/replica.rs#L522).

I'll spare you the debugging. It turns out what is happening is unrelated to `timerfd` or `async_timer::Interval`, and instead has to do with tasks that poll `io::Registration` and `io::Driver` to decide when to return `Pending`:

 1. The underlying fd becomes ready, and notifies the task, setting `NOTIFIED` in the task's state using [`transition_to_notified`](https://github.com/tokio-rs/tokio/blob/55b5e1b6adcc3f85a35d16444e496b203648c99d/tokio/src/task/state.rs#L246).
 2. The task is polled, which [clears the `NOTIFIED` bit](https://github.com/tokio-rs/tokio/blob/55b5e1b6adcc3f85a35d16444e496b203648c99d/tokio/src/task/harness.rs#L65).
 3. The task, when polled, eventually calls [`io::Registration::poll_ready`](https://github.com/tokio-rs/tokio/blob/55b5e1b6adcc3f85a35d16444e496b203648c99d/tokio/src/io/registration.rs#L215) to see if there's data available to it (e.g., from `Stream::poll_next`).
 4. `poll_ready` [registers the `Context`](https://github.com/tokio-rs/tokio/blob/55b5e1b6adcc3f85a35d16444e496b203648c99d/tokio/src/io/registration.rs#L228) for later notifications by calling [`io::Driver::register`](https://github.com/tokio-rs/tokio/blob/55b5e1b6adcc3f85a35d16444e496b203648c99d/tokio/src/io/driver/mod.rs#L274).
 5. `io::Driver::register` finds that the fd is ready as requested, and [wakes up the current task](https://github.com/tokio-rs/tokio/blob/55b5e1b6adcc3f85a35d16444e496b203648c99d/tokio/src/io/driver/mod.rs#L292) to ensure it gets polled again.
 6. `poll_ready` returns `Ok(Ready)` to indicate that the fd is ready to be read.
 7. The task does whatever it wanted to with the fd, and continues to call `poll_ready` until it returns `Ok(Pending)`, at which time the entire task returns `Pending`.
 8. When the task yields with `Pending`, the executor checks if the task [has been notified while it ran](https://github.com/tokio-rs/tokio/blob/55b5e1b6adcc3f85a35d16444e496b203648c99d/tokio/src/task/harness.rs#L140). If it was, it is [scheduled to run again](https://github.com/tokio-rs/tokio/blob/55b5e1b6adcc3f85a35d16444e496b203648c99d/tokio/src/task/harness.rs#L56). Since we woke the task up in 5, the task is scheduled to run again. This causes the second (spurious) call to `task.poll()`.

The issue, specifically, is that 5 issues a wakeup to the current task, even though we then _continue_ to call `poll_ready` until it returns `Pending`, at which point that wakeup becomes obsolete. The executor does not know this though, and `poll_ready` has no way to undo the wakeup, so the executor must re-execute the task.

#2221 contains a PR with a test that demonstrates this issue on `TcpListener.incoming`. I believe the issue is not restricted to `Stream`, and will occur for anything that calls `poll_ready` in a loop.

I'm not sure how we fix this? Presumably the wakeup in 5 is there for a reason in case the caller does _not_ continue to call `poll_ready` until it returns `Pending` (are they allowed to do that?). Another question is whether registering the `Context` in 4 is necessary if the waker hasn't changed and the fd has already been registered. If that's not the case, then we could augment that check so that we don't call `register`, and therefore don't incur the extra wakeup.

**Comment by jonhoo** [comment]
A fun observation, the notify call in 5 has been there since the very first tokio commit:
https://github.com/tokio-rs/tokio/blob/bc64194be17bc9711b4a56364e677d823c7cc3d1/src/event_loop.rs#L216-L225

It may well have predated the contract that you need to keep polling until you get `Pending`, and that's why it notifies preemptively (and unnecessarily under the contract). I'm not familiar enough with the internals to say exactly for sure though.

**Comment by jonhoo** [comment]
This was fixed by #2221!
